src/tablespace.c

```diff
@@ -0,0 +1,467 @@
+
+	/* Add the tablespace to the hypertable */
+	old = cache_switch_to_memory_context(hcache);
+	hypertable_add_tablespace(ht, tspcid, tspcoid);
+	MemoryContextSwitchTo(old);
+
+	cache_release(hcache);
+
+	PG_RETURN_VOID();
+}
+
+static int
+tablespace_detach_one(Oid hypertable_oid, const char *tspcname, Oid tspcoid)
+{
+	Cache	   *hcache;
+	Hypertable *ht;
+	int			ret;
+
+	hypertable_permissions_check(hypertable_oid, GetUserId());
+
+	hcache = hypertable_cache_pin();
+	ht = hypertable_cache_get_entry(hcache, hypertable_oid);
+
+	if (NULL == ht)
+		ereport(ERROR,
+				(errcode(ERRCODE_IO_HYPERTABLE_NOT_EXIST),
+				 errmsg("Table \"%s\" is not a hypertable",
+						get_rel_name(hypertable_oid))));
+
+	if (!hypertable_has_tablespace(ht, tspcoid))
+		ereport(ERROR,
+				(errcode(ERRCODE_IO_TABLESPACE_NOT_ATTACHED),
+			 errmsg("Tablespace \"%s\" is not attached to hypertable \"%s\"",
+					tspcname, get_rel_name(hypertable_oid))));
+
+	ret = tablespace_delete(ht->fd.id, tspcname);
+	hypertable_delete_tablespace(ht, tspcoid);
+
+	cache_release(hcache);
+
+	return ret;
+}
+
+static int
+tablespace_detach_all(Oid hypertable_oid)
+{
+	Cache	   *hcache;
+	Hypertable *ht;
+	int			ret;
+
+	hypertable_permissions_check(hypertable_oid, GetUserId());
+
+	hcache = hypertable_cache_pin();
+	ht = hypertable_cache_get_entry(hcache, hypertable_oid);
+
+	if (NULL == ht)
+		ereport(ERROR,
+				(errcode(ERRCODE_IO_HYPERTABLE_NOT_EXIST),
+				 errmsg("Table \"%s\" is not a hypertable",
+						get_rel_name(hypertable_oid))));
+
+	ret = tablespace_delete(ht->fd.id, NULL);
+
+	cache_release(hcache);
+
+	return ret;
+}
+
+TS_FUNCTION_INFO_V1(tablespace_detach);
+
+Datum
+tablespace_detach(PG_FUNCTION_ARGS)
+{
+	Oid			hypertable_oid = InvalidOid;
+	Name		tspcname;
+	Oid			tspcoid;
+	int			ret;
+
+	switch (PG_NARGS())
+	{
+		case 1:
+			tspcname = PG_GETARG_NAME(0);
+			break;
+		case 2:
+			hypertable_oid = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+			tspcname = PG_GETARG_NAME(1);
+			break;
+		default:
+			elog(ERROR, "Invalid number of arguments");
+	}
+
+	if (NULL == tspcname)
+		elog(ERROR, "Invalid tablespace name");
+
+	tspcoid = get_tablespace_oid(NameStr(*tspcname), true);
+
+	if (!OidIsValid(tspcoid))
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+				 errmsg("No tablespace \"%s\" exists.",
+						NameStr(*tspcname))));
+
+	if (OidIsValid(hypertable_oid))
+		ret = tablespace_detach_one(hypertable_oid, NameStr(*tspcname), tspcoid);
+	else
+		ret = tablespace_delete_from_all(NameStr(*tspcname), GetUserId());
+
+	PG_RETURN_INT32(ret);
+}
+
+TS_FUNCTION_INFO_V1(tablespace_detach_all_from_hypertable);
+
+Datum
+tablespace_detach_all_from_hypertable(PG_FUNCTION_ARGS)
+{
+	if (PG_NARGS() != 1)
+		elog(ERROR, "Invalid number of arguments");
+
+	if (PG_ARGISNULL(0))
+		elog(ERROR, "Invalid argument");
+
+	PG_RETURN_INT32(tablespace_detach_all(PG_GETARG_OID(0)));
+}
+
+TS_FUNCTION_INFO_V1(tablespace_show);
+
+Datum
+tablespace_show(PG_FUNCTION_ARGS)
```

Comment:
So when you call show_tablespaces() You may be asking two distinct things: 1) What tablespaces are attached to the current hypertable to be added to new chunks. 2) What are the tablespaces that have been assigned to the existing chunks of this hypertable. Right now this answers (1) not (2). Maybe we need to change the name to `show_assigned_tablespaces()` To make it more clear. Also do we need a function for (2).

Reply:
I figured `show_tablespaces()` is just a way to show you which tablespaces you've done `attach_tablespace()` on for a particular hypertable.

I think (2) is something entirely different and seems more akin to some debug or status function and is information you can get from the regular PG catalog.

Reply:
In that case I think `show_assigned_tablespaces` is more clear.

Reply:
What does assigned mean in this context? This function only shows the tablespaces that are attached to a hypertable. It does nothing w.r.t. which chunks are on which particular tablespaces, nor do I think it should.

Are you suggesting adding another function for showing how tablespaces are assigned to chunks?